content/blog/2025-05-dev-update/index.md

```diff
@@ -0,0 +1,100 @@
+  He also began writing [a new tool documentation set](https://github.com/tlaplus/tlaplus/pull/1181).
+  For the TLA⁺ VS Code Extension, he updated it to [incorporate level checking information from SANY](https://github.com/tlaplus/vscode-tlaplus/pull/367) and [exposed some APIs](https://github.com/tlaplus/vscode-tlaplus/pull/372) to be used by LLM agents.
+
+### TLA⁺ Foundation-funded updates
+
+Finally, things [Andrew Helwer](https://ahelwer.ca/) (author of this post) worked on - all funded by the TLA⁺ Foundation!
+- I spent a lot of time expanding the [*Create your own TLA⁺ tools*](https://docs.tlapl.us/creating:start) tutorial series, trying to encode as much knowledge about parsing TLA⁺ as possible.
+  As a highlight, the tutorial now has a module on parsing vertically-aligned conjunction & disjunction lists.
+  This is nearly complete; I've finished writing the code for two of the final three modules (operator parameters, actions, then safety checking) and now need to write them up.
+- I created & delivered [a talk](https://www.youtube.com/watch?v=KrhZebeRn90) at the TLA⁺ Community Event summarizing the current state of TLA⁺ development, what the TLA⁺ Foundation has been funding to improve it, and what I want to see in the near future.
+
+### Newbie Corner
+
+Here we highlight an issue that would be a good choice for new contributors!
+This is probably one of the most advanced issues yet profiled in this series, but it's very interesting.
+William Schultz [reported a bug](https://github.com/tlaplus/tlaplus/issues/220) where TLA⁺ cannot handle constraints like `<<y>>' = <<y>>`, which should be a synonym for `UNCHANGED y` or `y' = y`.
+
+Why is this bug interesting?
+When evaluating an action, TLC has to do a weird double-duty where it both evaluates the action as a simple logical predicate over variables - evaluating to true or false - while also extracting hints about the possible values of those same variables!
+Consider what happens when TLC sees an action like:
+```
+Action ==
+  /\ x = 0
+  /\ x' = x + 1
+```
+When it evaluates the expression `x' = x + 1`, it *doesn't yet know the value of `x'`*!
+It has to slyly recognize that the as-yet-undefined value of `x'` is on the left hand side of an equals operator, so can be assigned/constrained to the value `x + 1`; then, having defined the value of `x'`, it evaluates `x' = x + 1` as an ordinary expression which will of course be true.
+
+Wrapping your head around this weird double-duty of TLC's interpreter will bring you great understanding of TLA⁺ tool internals.
+In TLA⁺, it isn't only variables that can be primed - entire expressions can be primed.
+When that happens, it is the same as individually priming all the variables within that expression.
+So, `<<y>>'` is the same as `<<y'>>`.
+
+Fixing this issue requires looking at how TLC evaluates the equals operator when both parameters are tuples, and more specifically when the left-hand-side is primed.
```

Comment:
I'm not convinced that the proposed solution is optimal. At the very least, this call to action should raise awareness of potential performance implications. TLC omits support for certain expressions due to its optimization for speed.

The idea of a pre-processing stage has been discussed multiple times, and this is a case where the expression could be desugared to `x' = x` or `UNCHANGED x`.  Note that Apalache has a desugaring stage.